projects/cm_mcu/MonitorI2CTask.c

```diff
@@ -0,0 +1,195 @@
+
+#define TMPBUFFER_SZ 96
+
+// Monitor registers of FF temperatures, voltages, currents, and ClK statuses via I2C
+void MonitorI2CTask(void *parameters)
+{
+
+  struct MonitorI2CTaskArgs_t *args = parameters;
+
+  configASSERT(args->name != 0);
+
+  // watchdog info
+  task_watchdog_register_task(kWatchdogTaskID_MonitorI2CTask);
+
+  // initialize to the current tick time
+  args->updateTick = xTaskGetTickCount();
+
+  // wait for the power to come up
+  vTaskDelayUntil(&(args->updateTick), pdMS_TO_TICKS(2500));
+
+  int IsCLK = (strstr(args->name, "CLK") != NULL);   // the instance is of CLK-device type
+  int IsFF12 = (strstr(args->name, "FF12") != NULL); // the instance is of FF 12-ch part type
+  // int IsFFDAQ =  (strstr(args->name, "FFDAQ") != NULL);  //the instance is of FF 4-ch part type (DAQ links) -- not being used currently
+
+
+  // reset the wake time to account for the time spent in any work in i2c tasks
+
+  bool good = false;
+  for (;;) {
+    char tmp[TMPBUFFER_SZ];
+
+    // grab the semaphore to ensure unique access to I2C controller
+    while (xSemaphoreTake(args->xSem, (TickType_t)10) == pdFALSE)
+        ;
+    args->updateTick = xTaskGetTickCount(); // current time in ticks
+    // -------------------------------
+    // loop over devices in the device-type instance
+    // -------------------------------
+    for (int ps = 0; ps < args->n_devices; ++ps) {
+
+      if (!IsCLK) { // Fireflies need to be checked if the links are connected or not
+        int offsetFF12 = 1 - IsFF12;
+        if (!isEnabledFF(ps + (offsetFF12 * (NFIREFLIES_IT_F1)) + ((args->i2c_dev - I2C_DEVICE_F1) * (-1) * (NFIREFLIES_F2)))) // skip the FF if it's not enabled via the FF config
+          continue;
+        args->updateTick = xTaskGetTickCount();
+      }
+
+      if (args->requirePower) {
+        if (getPowerControlState() != POWER_ON) {
+          if (good) {
+            snprintf(tmp, TMPBUFFER_SZ, "MONI2C(%s): 3V3 died. Skipping I2C monitoring.\r\n", args->name);
+            log_info(LOG_MONI2C, "%s: PWR off. Disabling I2C monitoring.\r\n", args->name);
+            good = false;
+            task_watchdog_unregister_task(kWatchdogTaskID_MonitorI2CTask);
+          }
+          vTaskDelayUntil(&(args->updateTick), pdMS_TO_TICKS(500));
+          continue;
+        }
+        else if (getPowerControlState() == POWER_ON) { // power is on, and ...
+          if (!good) {                                 // ... was not good, but is now good
+            task_watchdog_register_task(kWatchdogTaskID_MonitorI2CTask);
+            snprintf(tmp, TMPBUFFER_SZ, "MONI2C(%s): 3V3 came back. Restarting I2C monitoring.\r\n", args->name);
+            log_info(LOG_MONI2C, "%s: PWR on. (Re)starting I2C monitoring.\r\n", args->name);
+            good = true;
+          }
+        }
+        // if the power state is unknown, don't do anything
+      }
+
+      // select the appropriate output for the mux
+      uint8_t data[1];
```

Comment:
it's not very idiomatic in C to have an array of length 1. why use an array?

Reply:
It might be an artifact from following MONtask. change it to one-byte variable. Done.